include/RAJA/policy/cuda/forall.hpp

```diff
@@ -74,22 +74,40 @@
 
 #include "RAJA/index/IndexSet.hpp"
 
+#include <algorithm>
+
 namespace RAJA
 {
 
+namespace cuda
+{
+
 namespace impl
 {
-//
-//////////////////////////////////////////////////////////////////////
-//
-// CUDA kernel templates.
-//
-//////////////////////////////////////////////////////////////////////
-//
 
-// INTERNAL namespace to encapsulate helper functions
-namespace INTERNAL
+/*!
+ ******************************************************************************
+ *
+ * \brief calculate gridDim from length of iteration and blockDim
+ *
+ ******************************************************************************
+ */
+RAJA_INLINE
+dim3 getGridDim(size_t len, dim3 blockDim)
 {
+  size_t block_size = blockDim.x * blockDim.y * blockDim.z;
+
+  int device = 0;
+  static int num_sm[8] {0, 0, 0, 0, 0, 0, 0, 0};
+  if (num_sm[device] == 0) {
+    cudaDeviceGetAttribute(&num_sm[device], cudaDevAttrMultiProcessorCount, device);
+  }
+  size_t max_concurrent_blocks = num_sm[device] * (MAX_THREADS_PER_SM/block_size) ;
+
+  size_t gridSizeMax = (len + block_size-1) / block_size;
+
+  return std::min(gridSizeMax, max_concurrent_blocks);
```

Comment:
@trws This all sounds good. Since, we're using CUB as the performance target it's definitely reasonable to tweak this calculation, and besides I'm curious about the oversubscription factor, or anything else that we leverage. Another interesting exercise is to see whether we can gain an advantage with multiple reducers in the same kernel.